force-app/main/default/classes/OpportunityHelper.cls

```diff
@@ -0,0 +1,240 @@
+public with sharing class OpportunityHelper {
+
+    public virtual class BaseException extends Exception {}
+    public class OtherException extends BaseException {}    
+
+    public static void checkOpportunityAmount(List<Opportunity> newOppList){
+
+        for(Opportunity opp : newOppList){
+            if(opp.Amount <= 5000){
+                opp.addError('Opportunity amount must be greater than 5000');
+            }
+        }
+    
+    }
+
+    public static void setPrimaryContact(List<Opportunity> newOppList){
+
+        // Check for Contact read and create permissions
+        if (!Schema.sObjectType.Contact.isAccessible()) {
+            throw new OtherException('Current user does not have permission to read or create Contact records.');
+        }
+
+        Set<Id> accountIdSet = new Set<Id>();
+
+        for(Opportunity opp : newOppList){
+            
+            if(opp.AccountId != null){
+                accountIdSet.add(opp.AccountId);
+            }
+
+        }
+
+        List<Contact> contactList = [
+            SELECT
+                Id,
+                AccountId,
+                Name
+            FROM
+                Contact
+            WHERE
+                AccountId IN :accountIdSet
+                AND Title = 'CEO'
+        ];
+
+        Map<Id, Id> accountToContactMap = new Map<Id,Id>();
+
+        for(Contact contactRecord : contactList){
+            accountToContactMap.put(contactRecord.AccountId, contactRecord.Id);
+        }
+
+        for(Opportunity opportunityRecord : newOppList){
+            if (opportunityRecord.Primary_Contact__c == null) {
+                opportunityRecord.Primary_Contact__c = accountToContactMap.get(opportunityRecord.AccountId);
+            }
+        }
+ 
+    }
+
+    public static void validateDeleteCriteria(List<Opportunity> oldOppList){
+
+        // Check for Account read and create permissions
+        if (!Schema.sObjectType.Account.isAccessible()) {
+            throw new OtherException('Current user does not have permission to read Account records.');
+        }
+
+        // Step01: Collect Account Ids
+        Set<Id> accountIdSet = new Set<Id>();
+
+        for(Opportunity opp : oldOppList){
+            if(opp.AccountId != null){
+                accountIdSet.add(opp.AccountId);
+            }
+        }
+
+        // Step02: Query Account records
+        Map<Id, Account> accountMap = new Map<Id, Account>(
+            [Select Id, Name, Industry from Account where Id IN :accountIdSet]
+        );
+
+        // Step03: Process each Opportunity
+        for(Opportunity opp : oldOppList) {
+            if(opp.AccountId != null
+                && accountMap.containsKey(opp.AccountId)){
+                Account relatedAccount = accountMap.get(opp.AccountId);
+                String accountIndustry = relatedAccount.Industry;
+                // Prevent delete if the industry is Banking and opportunity is closed won
```

Comment:
You might be able to combine these lines into a more efficient query, try it and let me know if it works

`  

            Account acc = new Account();

           Map<Id, Account> accounts = new Map<Id, Account>(
            [SELECT Id, Industry 
            FROM Account 
            WHERE Id IN (
                SELECT AccountId 
                FROM Opportunity 
                WHERE Id IN :oldOppList)
            ]);

        for(Opportunity opp : oldOppList){
            acc = accounts.get(opp.AccountId);
`